napari/_qt/qt_main_window.py

```diff
@@ -586,6 +586,48 @@ def remove_dock_widget(self, widget):
         else:
             self._qt_window.removeDockWidget(widget)
 
+    def add_magic_function(self, function, magic=None, dock=None):
```

Comment:
I'm open to both changes, curious what @jni and others think

Reply:
> (at least, I can't think of a scenario with this code where that would lead to unexpected results, @jni?)

Reading the body of the function, I agree, but literally yesterday, for a tiny utility function in an interactive session, for the same issue, I was like, "nah, I don't need to worry about it here" and sure enough it exploded in my face. 😂 So I think we should stick to accepted good practice and keep the Nones, even if it's a bit uglier.

I *can* imagine future contributors coming in here and messing with the dictionary and breaking things unexpectedly.

> I wonder about naming it add_function_widget rather than add_magic_function? Says a little more about what it does rather than how it does it. (but don't feel strongly)

Haha "how it does it"... "with magic!" 😂 I like `add_function_widget` quite a bit more than `add_magic_function`. I presume we envision this working with both magical and muggle functions. 😜 

> I'd like to hear people's thoughts on making magic and dock just **kwargs... and then pulling out the appropriate arguments for each internal function. would allow

One issue is whether we envision magicgui and dock kwargs ever clashing. It's not very far-fetched, is it? One suggestion: do both? Allow `**kwargs` with smart unpacking, *and* add two parameters, `magicgui_kwargs` and `dock_kwargs`, where people who like being explicit, or, in the case of clashes, people who *need* to be explicit, can do so.

Reply:
> One issue is whether we envision magicgui and dock kwargs ever clashing. It's not very far-fetched, is it? One suggestion: do both? Allow **kwargs with smart unpacking, and add two parameters, magicgui_kwargs and dock_kwargs, where people who like being explicit, or, in the case of clashes, people who need to be explicit, can do so.

hmm, I'm a little unsure about opening up both, if what about just doing one, so `magic` was the magicgui dict, which is maybe more involved, and then there was no `dock` keyword arg which is maybe a bit weird, but we just had the same kwargs as the `add_dock_widget`?

I think that might be quite nice actually. The API for `add_function_widget` and `add_dock_widget` are now basically the same apart from the function one takes a function instead of a widget directly, and the function one takes an additional kwarg `magic`.

I feel like if we start here and people want to just pass magic gui kwargs directly we could add that in the future. Thoughts @tlambert03 @jni?

Reply:
> I feel like if we start here and people want to just pass magic gui kwargs directly we could add that in the future. Thoughts @tlambert03 @jni?

I think what we do now is probably how it's "going to be" (once examples get out there, they don't tend to change easily)... :) but as I said, I don't feel too strongly about it either way.  I'm also ok with your proposal for `magic` or `magic_kwargs` being a dict, and then having all the standard dock_widget kwargs being spelled out.  your call

Reply:
Dunno... I think I'd be more interested in passing magicgui kwargs, since they are harder (impossible?) to modify after instantiation... Whereas moving a dockwidget around is fine.

I think the least controversial choice here is the two kwarg dicts. (Though I do think they should have the word `kwargs` in the argument name — that is standard.) I suggest we start with that and open a followup issue for this discussion.

Reply:
actually... I am swayed by @jni's point about potential clashes, particularly in the magic_gui args (which, if I'm not mistaken, can be scarily flexible since they can be named after parameter args)... so let's _not_ do what I originally suggested for magicgui **kwargs.  I'm fine with `dock_kwargs` and `magic_kwargs`.

> they are harder (impossible?) to modify after instantiation
fwiw: after the "mythical" magicgui rewrite that will merge someday... modifying magic kwargs after instantiation is quite straightforward (though the range of what's possible is going to be restricted slightly at first)

Reply:
have renamed `add_function_widget` and gone with `magic_kwargs` and then individual kwargs for dock widget. I think there's quite nice the symmetry now between the `add_dock_widget` and `add_function_widget` API.

Reply:
renamed and addressed